forge/src/main/java/me/lucko/luckperms/forge/model/ForgeUser.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of LuckPerms, licensed under the MIT License.
+ *
+ *  Copyright (c) lucko (Luck) <luck@lucko.me>
+ *  Copyright (c) contributors
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge.model;
+
+import me.lucko.luckperms.common.cacheddata.type.MetaCache;
+import me.lucko.luckperms.common.cacheddata.type.PermissionCache;
+import me.lucko.luckperms.common.context.manager.QueryOptionsCache;
+import me.lucko.luckperms.common.locale.TranslationManager;
+import me.lucko.luckperms.common.model.User;
+import me.lucko.luckperms.common.verbose.event.CheckOrigin;
+import net.luckperms.api.query.QueryOptions;
+import net.luckperms.api.util.Tristate;
+import net.minecraft.server.level.ServerPlayer;
+
+import java.util.Locale;
+import java.util.function.Function;
+
+public class ForgeUser {
```

Comment:
Sorry.. I know we removed Mixins, but this seems like a good candidate... what do you think?

Failing that, the other option would be to use the InlineQueryOptionsSupplier, like BungeeCord does: https://github.com/LuckPerms/LuckPerms/blob/master/bungee/src/main/java/me/lucko/luckperms/bungee/context/BungeeContextManager.java#L64

I'd prefer that as opposed to having an additional map of users in the context manager

Reply:
I've switched to using Forge Capabilities https://github.com/LuckPerms/LuckPerms/pull/3262/commits/21f3805cf3355cb346cf2e87dbb4af6d645f8597 the `UserCapability` (previously `ForgeUser`) is stored on the `ServerPlayer` object without the need for Mixins.

Reply:
Perfect :+1: